docs/build/voting-dapp-tutorial.md

```diff
@@ -0,0 +1,1245 @@
+```clojure
+"Loading election.repl..."
+"Begin Tx 0"
+"Setting transaction keys"
+"Setting transaction data"
+"Loading election.pact..."
+"Keyset defined"
+"Loaded module election, hash jOBJiSEH5HVsAmLNSdHq_D3Kl6TkFpCJcEWPtJD86Gc"
+"Voted A!"
+"Commit Tx 0"
+```
+
+Let's continue by adding a bit more logic. Copy this snippet just below the module declaration:
+
+```clojure
+  (defschema candidates-schema
+    "Candidates table schema"
+    name:string
+    votes:integer)
+
+  (defschema votes-schema
+    "Votes table schema"
+    cid:string
+  )
+
+  (deftable votes:{votes-schema})
+
+  (deftable candidates:{candidates-schema})
+  ```
+In the snippet we've got several definitions:
+* `defschema candidates-schema` -> defines the schema for our **candidates** table where we will keep track of the candidates that users can vote for
+* `defschema votes-schema` -> defines the schema for our **votes** table where we will store each vote
+* `deftable votes:{votes-schema}` -> defines the `votes` table that will use the schema defined above
+* `deftable candidates:{candidates-schema}` -> defines the `candidates` table that will use the schema defined above
+
+:::note
+Pact implements a key-row model which means a row is accessed by a single key. It is our responsibility as developers to design the schema in a way that we can retrieve the information that we need using a single row query. Multiple row queries are expensive which we will see in an example a bit later.
+:::
+
+Let's continue by implementing our `vote` function:
+
+```clojure
+  (defun vote (account:string cid:string)
+    "Submit a new vote"
+
+    (let ((doubleVote (user-voted account)))
+      (enforce (= doubleVote false) "Multiple voting not allowed"))
+
+    (let ((exists (candidate-exists cid)))
+      (enforce (= exists true) "Candidate doesn't exist"))
+
+    (with-capability (ACCOUNT-OWNER account)
+      (vote-protected account cid))
+
+    (format "Voted for candidate {}!" [cid])
+  )
+```
+
+Quite a lot happening here, so let's go line by line:
+* to prevent double voting, we have to check if the user already voted. Below is the implementation of the `user-voted` function which does that.
+```clojure
+(defun user-voted:bool (account:string)
+  (with-default-read votes account
+    { "cid": "" }
+    { "cid":= cid }
+    (> (length cid) 0))
+)
+```
+* to prevent voting for a candidate that doesn't exist we implemented `candidate-exists` function and we `enforce` it returns `true` for our parameters
+
+```clojure
+(defun candidate-exists:bool (cid:string)
+  (with-default-read candidates cid
+    { "name": "" }
+    { "name" := name }
+    (> (length name) 0))
+)
+```
+
+:::note
+Notice how we used a similar approach in both `user-voted` and `candidate-exists` functions to execute a single row read using `with-default-read`. Below is an alternative, naive a way to implement the same thing:
+
+```clojure
+(contains account (keys votes))
```

Comment:
Would rather not even mention `keys` here. Just pretend like it doesn't exist

Reply:
My opinion here is that is very likely people will use `keys` and "how not to do it" example is useful. But then yes, you have them wondering what's the purpose of `keys`.

Reply:
I removed that example. I fear that many will still use this function because it's in the docs.